[REVIEW] Proposal for guidelines on stateful and stateless cuML APIs [skip ci] (#672)
rapidsai/cuml: cpp/DEVELOPER_GUIDE.md

  void fit(const T *input, int n_rows, int n_cols, const int *labels);
  void predict(const T *input, int n_rows, int n_cols, int *predictions);
};

void decisionTreeClassifierFit(const cumlHandle &handle, const float *input, int n_rows, int n_cols,
                               const int *labels, DecisionTreeClassifier<float> *model, DTParams params,
                               bool verbose=false);
void decisionTreeClassifierPredict(const cumlHandle &handle, const float* input,
                                   DecisionTreeClassifier<float> *model, int n_rows,
                                   int n_cols, int* predictions, bool verbose=false);
```

One of the right ways to expose this interface from `libcuml++.so` is:
```cpp
// NOTE: this example assumes that TreeNode and DTParams are the model/state that need to be stored
// and passed between fit and predict methods
template <typename T> struct TreeNode { /* nested tree-like data structure, but written as a POD! */ };
struct DTParams { /* hyper-params for building DT */ };

void decisionTreeClassifierFit(const cumlHandle &handle, const float *input, int n_rows, int n_cols,

>                                const int *labels, TreeNode<float> *&root, DTParams params,

Reviewer: Sorry for the naive question.
If the pyhton wrapper passes TreeNode<float> *root = NULL and the c++ code populates the recursive tree structure. How does that work ? Can the python API traverse through a c++ populated tree.

Author: If we need to traverse/save/load the tree from python, we could expose those methods in cpp api itself and the wrap those with necessary cython calls. What say?

Author: I didn't put those things in the example since that'd have made the example itself to be more complex

Author: How about now @vishalmehta1991? I have added a small section to just explain this.

The implementation of cuML is single threaded.

## Managing state
### Introduction
Every ML algo needs to store some state, eg: model and its related hyper-parameters. Thus, this section lays out guidelines for managing state along the API of cuML.

### Inside `libcuml++.so` aka our C++ interface

> Functions exposed via the cuML-C++ layer must be stateless. Meaning, they must accept all the required inputs, parameters and outputs in their argument list only. In other words, the stateful API should always be a wrapper around the stateless methods, NEVER the other way around. That said, internally, these stateless functions are free to create their own temporary classes, as long as they are not exposed on the interface of `libcuml++.so`.

Reviewer 2: "[...] in their argument list only" => Maybe add ", which should match the requirements listed below." 

"In other words, the stateful API should always be a wrapper around [...]" => "The stateful API, if one exists, should always be a wrapper around [...]" and/or link to the **scikit-learn-esq stateful API in C++** section below, so a first time reader can see which stateful API we're referring to.

Author: Thanks. Done.

Things which are OK to be exposed on the interface:

> 1. Any [POD](https://en.wikipedia.org/wiki/Passive_data_structure).

Reviewer 2: Would it be useful for developers if we also mentioned `std::is_pod` (c++11) as a way to check?

Author: Makes sense. Updated.

2. `cumlHandle` - since it stores GPU-related state which has nothing to do with the model/algo state.
3. Pointers (explicitly putting it out, even though can be considered as a POD).

> Taking decisiontree-classifier algo as an example, the following way of exposing its API would be wrong according to the guidelines in this section. Because, the we have exposed a C++ class with state along the interface of `libcuml++.so`.

Reviewer 2: Minor typo:
"Because, the we have exposed a C++ class with state along the interface of `libcuml++.so`."  => Maybe change to something like:
"This API exposes a non POD C++ class object, with state, along the interface of `libcuml++.so`."

Author: Thanks. Done.

```cpp
template <typename T>
class DecisionTreeClassifier {
  TreeNode<T>* root;
  DTParams params;
  const cumlHandle &handle;
public:
  DecisionTreeClassifier(const cumlHandle &handle, DTParams& params, bool verbose=false);
  void fit(const T *input, int n_rows, int n_cols, const int *labels, unsigned int *rowids,
           int n_sampled_rows, int unique_labels);
  void predict(const T *input, int n_rows, int n_cols, int *predictions);
};

> void decisionTreeClassifierFit(const cumlHandle &handle, const float *input, int n_rows, int n_cols,

Reviewer 2: If there is no expectation of having the example fully match the API of an existing ML algorithm, then maybe we could delete `rowids`, `n_sampled_rows`, `unique_labels` from `fit`, so the extra information does not distract the reader from what's at the core of the example? Same comment for the updated API.

Author: Good point. Thanks for catching this. Fixed.

The implementation of cuML is single threaded.

> ## Managing state

Reviewer 3: I think this strongly relates to the existing `## C APIs` section. So this should either be a subsection of `## C APIs` should be merged into this.

Author: Hmm.. I think the title of the section can cause confusion among the readers. This section is mostly to deal with how to expose state-related info from the __C++__ interface. Let me rename the section properly.

### Introduction
Every ML algo needs to store some state, eg: model and its related hyper-parameters. Thus, this section lays out guidelines for managing state along the API of cuML.

### Inside `libcuml++.so` aka our C++ interface

> Functions exposed via the cuML-C++ layer must be stateless. Meaning, they must accept all the required inputs, parameters and outputs in their argument list only, which should meet the requirements listed below. In other words, the [stateful API](#scikit-learn-esq-stateful-api-in-c) should always be a wrapper around the stateless methods, NEVER the other way around. That said, internally, these stateless functions are free to create their own temporary classes, as long as they are not exposed on the interface of `libcuml++.so`.

Reviewer 3: I think it would be good to the ability to provide C language bindings, for calls from FORTRAN and other languages, as a motivation to this section.

Things which are OK to be exposed on the interface:
1. Any [POD](https://en.wikipedia.org/wiki/Passive_data_structure) - one can use [std::is_pod](https://en.cppreference.com/w/cpp/types/is_pod) in C++11 to check POD types.
2. `cumlHandle` - since it stores GPU-related state which has nothing to do with the model/algo state.

> 3. Pointers (explicitly putting it out, even though can be considered as a POD).

Reviewer 3: To my understanding we can't allow pointers to non POD types as that will no longer allow us to provide C bindings. So maybe better `Pointer to POD types`?

Author: That's a good point. Didn't realize this statement could be interpreted that way! Let me update this.

Author: DOne.

```cpp
template <typename T>
class DecisionTreeClassifier {
  TreeNode<T>* root;
  DTParams params;
  const cumlHandle &handle;
public:
  DecisionTreeClassifier(const cumlHandle &handle, DTParams& params, bool verbose=false);
  void fit(const T *input, int n_rows, int n_cols, const int *labels);
  void predict(const T *input, int n_rows, int n_cols, int *predictions);
};

void decisionTreeClassifierFit(const cumlHandle &handle, const float *input, int n_rows, int n_cols,
                               const int *labels, DecisionTreeClassifier<float> *model, DTParams params,
                               bool verbose=false);
void decisionTreeClassifierPredict(const cumlHandle &handle, const float* input,
                                   DecisionTreeClassifier<float> *model, int n_rows,
                                   int n_cols, int* predictions, bool verbose=false);
```

One of the right ways to expose this interface from `libcuml++.so` is:
```cpp
// NOTE: this example assumes that TreeNode and DTParams are the model/state that need to be stored
// and passed between fit and predict methods

> template <typename T> struct TreeNode { /* nested tree-like data structure, but written as a POD! */ };

Reviewer 3: I must admit I did not spend to much thought on this, but how to use this type with a C binding?

The implementation of cuML is single threaded.

## Public cuML interface
### Motivation

> Such a well-defined interface is useful for two main reasons:

Reviewer 3: What about: `The public cuML interface is stateless because:` instead of `Such a well-defined interface is useful for two main resons:`

Author: Done.

1. Easily pickle ML algo's state information (model, hyper-params, etc), at the python layer
2. Provide a proper C-binding to interface with languages that don't understand C++
Thus, this section lays out guidelines for managing state along the API of cuML.

### General guideline
Functions exposed via the base cuML-c++ layer must be stateless. Things that are OK to be expoesd on the interface:
1. Any [POD](https://en.wikipedia.org/wiki/Passive_data_structure) - one can use [std::is_pod](https://en.cppreference.com/w/cpp/types/is_pod) in C++11 to check POD types.

> 2. `cumlHandle` - since it stores GPU-related state which has nothing to do with the model/algo state. If you're working on a C-binding, use `cumlHandle_t`, instead.

Reviewer 3: I would suggest to add a reference to `cpp/src/cuML_api.h` where the C API to manage `cumlHandle_t` is defined.

Author: Done.

```

One of the right ways to expose this would be:
```cpp
// NOTE: this example assumes that TreeNode and DTParams are the model/state that need to be stored
// and passed between fit and predict methods
template <typename T> struct TreeNode { /* nested tree-like data structure, but written as a POD! */ };
struct DTParams { /* hyper-params for building DT */ };
typedef TreeNode<float> TreeNodeF;
typedef TreeNode<double> TreeNodeD;

void decisionTreeClassifierFit(const cumlHandle &handle, const float *input, int n_rows, int n_cols,
                               const int *labels, TreeNodeF *&root, DTParams params,
                               bool verbose=false);
void decisionTreeClassifierPredict(const cumlHandle &handle, const double* input, int n_rows,
                                   int n_cols, const TreeNodeD *root, int* predictions,
                                   bool verbose=false);
```
The above example obviously under-plays the complexity involved with exposing a tree-like data structure across the interface! However, this example should be simple enough to drive the point across.

> ### WIP TODO (Need to explore this further!)

Reviewer 3: I think it is worth considering to drop this section from the developer guide. IMHO it adds too much details making the guide too verbose.

Author: @vishalmehta1991 preferred to keep such examples.

Reviewer 4: I think it is very useful to keep this section here, otherwise I can foresee receiving a PR where we need to request the store/load methods

Reviewer 3: I am just concerned that a too verbose developer guide will not be read. In general I think we should try to keep it concise.

Author: I agree that a concise guide should be our aim. But I think we should keep this section for now and depending on how it gets received, we shall decide whether to keep it or not. What say?

Reviewer 3: Fine with me.

The above example also means, it is the responsibility of C++ API to expose methods to load and store (aka marshalling) such a data structure. Further continuing this example, we could expose the following methods to achieve this:
```cpp
void storeTreeNode(const TreeNodeF *root, std::ostream &os);
void storeTreeNode(const TreeNodeD *root, std::ostream &os);
void loadTreeNode(TreeNodeF *&root, std::istream &is);
void loadTreeNode(TreeNodeD *&root, std::istream &is);
```
It is also worthy to note that for algos like GLM, where the model consists of an array of weights, such a custom load/store methods are not explicitly needed.

### `libcuml.so` (aka C-API wrapper over C++ API)

> As long as the guidelines in the above sub-sections have been followed, it must be easy to C-wrap the C++ API. Refer to dbscan as an example on how to properly wrap the C++ API with a C-binding. In short:

Reviewer 3: I would suggest to add a link to `cpp/src/dbscan/dbscan_api.h` and `cpp/src/dbscan/dbscan_api.cpp` here.

Author: Done.

The implementation of cuML is single threaded.

## Public cuML interface
### Motivation

> Public cuML interface is stateless for two main reasons:

Reviewer 4: ```suggestion
The public cuML interface is stateless for two main reasons:
```

> 1. Easily pickle ML algo's state information (model, hyper-params, etc), at the python layer

Reviewer 4: ```suggestion
1. To ease the serialization of ML algorithm's state information (model, hyper-params, etc), enabling features such as easy pickling in the python layer.
```

> 2. Provide a proper C-binding to interface with languages that don't understand C++

Reviewer 4: ```suggestion
2. To provide proper C bindings for interfacing with languages that can't consume C++ APIs  directly.
```

> Thus, this section lays out guidelines for managing state along the API of cuML.

Author: Done

### General guideline

> As mentioned before, functions exposed via the base cuML-c++ layer must be stateless. Things that are OK to be expoesd on the interface:

Reviewer 4: ```suggestion
As mentioned before, functions exposed via the base cuML C++ layer must be stateless. Things that are OK to be exposed on the interface:
```

> 1. Any [POD](https://en.wikipedia.org/wiki/Passive_data_structure) - one can use [std::is_pod](https://en.cppreference.com/w/cpp/types/is_pod) in C++11 to check POD types.

Reviewer 4: ```suggestion
1. Any [POD](https://en.wikipedia.org/wiki/Passive_data_structure) - see [std::is_pod](https://en.cppreference.com/w/cpp/types/is_pod) as a reference for C++11  POD types.
```

2. `cumlHandle` - since it stores GPU-related state which has nothing to do with the model/algo state. If you're working on a C-binding, use `cumlHandle_t`([reference](src/cuML_api.h)), instead.
3. Pointers to POD types (explicitly putting it out, even though can be considered as a POD).

> Internally, for the C++ base layer atleast, these stateless functions are free to use their own temporary classes, as long as they are not exposed on the interface.

Reviewer 4: ```suggestion
Internally, for the C++ base layer at least, these stateless functions are free to use their own temporary classes, as long as they are not exposed on the interface.
```

### `libcuml++.so` (aka C++ API)

> Taking decisiontree-classifier algo as an example, the following way of exposing its API would be wrong according to the guidelines in this section. Because, this API exposes a non-POD C++ class object along the C++ API.

Reviewer 4: ```suggestion
Using the Decision Tree Classifier algorithm as an example, the following way of exposing its API would be wrong according to the guidelines in this section, since it exposes a non-POD C++ class object in the C++ API:
```

```cpp
template <typename T>
class DecisionTreeClassifier {
  TreeNode<T>* root;
  DTParams params;
  const cumlHandle &handle;
public:
  DecisionTreeClassifier(const cumlHandle &handle, DTParams& params, bool verbose=false);
  void fit(const T *input, int n_rows, int n_cols, const int *labels);
  void predict(const T *input, int n_rows, int n_cols, int *predictions);
};

void decisionTreeClassifierFit(const cumlHandle &handle, const float *input, int n_rows, int n_cols,
                               const int *labels, DecisionTreeClassifier<float> *model, DTParams params,
                               bool verbose=false);
void decisionTreeClassifierPredict(const cumlHandle &handle, const float* input,
                                   DecisionTreeClassifier<float> *model, int n_rows,
                                   int n_cols, int* predictions, bool verbose=false);
```

> One of the right ways to expose this would be:

Reviewer 4: ```suggestion
An alternative correct way to expose this could be:
```

```cpp
// NOTE: this example assumes that TreeNode and DTParams are the model/state that need to be stored
// and passed between fit and predict methods
template <typename T> struct TreeNode { /* nested tree-like data structure, but written as a POD! */ };
struct DTParams { /* hyper-params for building DT */ };
typedef TreeNode<float> TreeNodeF;
typedef TreeNode<double> TreeNodeD;

void decisionTreeClassifierFit(const cumlHandle &handle, const float *input, int n_rows, int n_cols,
                               const int *labels, TreeNodeF *&root, DTParams params,
                               bool verbose=false);
void decisionTreeClassifierPredict(const cumlHandle &handle, const double* input, int n_rows,
                                   int n_cols, const TreeNodeD *root, int* predictions,
                                   bool verbose=false);
```

> The above example obviously under-plays the complexity involved with exposing a tree-like data structure across the interface! However, this example should be simple enough to drive the point across.

Reviewer 4: ```suggestion
The above example understates the complexity involved with exposing a tree-like data structure across the interface! However, this example should be simple enough to drive the point across.
```

### WIP TODO (Need to explore this further!)

> The above example also means, it is the responsibility of C++ API to expose methods to load and store (aka marshalling) such a data structure. Further continuing this example, we could expose the following methods to achieve this:

Reviewer 4: ```suggestion
This guidelines also mean that it is the responsibility of C++ API to expose methods to load and store (aka marshalling) such a data structure. Further continuing the Random Forest example,  the following methods could achieve this:
```

```cpp

> void storeTreeNode(const TreeNodeF *root, std::ostream &os);

Reviewer 3: Wondering if the naming of the function is as intended. Does `storeTreeNode` store the whole tree starting at `root`? If so I think it should be named `storeTree`.

Author: Done

void storeTreeNode(const TreeNodeD *root, std::ostream &os);
void loadTreeNode(TreeNodeF *&root, std::istream &is);
void loadTreeNode(TreeNodeD *&root, std::istream &is);
```

> It is also worthy to note that for algos like GLM, where the model consists of an array of weights, such a custom load/store methods are not explicitly needed.

Reviewer 4: ```suggestion
It is also worth noting that for algorithms such as the members of GLM, where models consist of an array of weights and therefore easy to manipulate directly by the users, such custom load/store methods might not explicitly needed.
```

### `libcuml.so` (aka C-API wrapper over C++ API)

> As long as the guidelines in the above sub-sections have been followed, it must be easy to C-wrap the C++ API. Refer to [dbscan](src/dbscan/dbscan_api.h) as an example on how to properly wrap the C++ API with a C-binding. In short:

Reviewer 4: ```suggestion
Following the guidelines outlined above will ease the process of "C-wrapping" the C++ API. Refer to [DBSCAN](src/dbscan/dbscan_api.h) as an example on how to properly wrap the C++ API with a C-binding. In short:
```

1. Use only C compatible types or objects that can be passed as opaque handles (like `cumlHandle_t`).
2. Using templates is fine if those can be instantiated from a specialized C++ function with `extern "C"` linkage.

> 3. Expose custom create/load/store/destroy methods, if the model is more complex than an array of parameters (eg: RF). One possible way of working with such exposed states from C++ layer is shown in the sample repo [here](https://github.com/teju85/managing-state-cuml).

Reviewer 4: ```suggestion
3. Expose custom create/load/store/destroy methods, if the model is more complex than an array of parameters (eg: Random Forest). One possible way of working with such exposed states from the C++ layer is shown in a sample repo [here](https://github.com/teju85/managing-state-cuml).
```

### scikit-learn-esq stateful API in C++

> This API should always be a wrapper around the stateless methods, NEVER the other way around. We are [still discussing](https://github.com/rapidsai/cuml/issues/456) about the right way to expose such a wrapper API around `libcuml++.so`. So, stay tuned for more details.

Reviewer 4: ```suggestion
This API should always be a wrapper around the stateless methods, NEVER the other way around. The design discussion about the right way to expose such a wrapper API around `libcuml++.so` is [still going on](https://github.com/rapidsai/cuml/issues/456)  So, stay tuned for more details.
```

### File naming convention

> 1. An ML algorithm `<algo>` is to be implemented inside the folder named `src/<algo>`.

Reviewer 4: ```suggestion
1. An ML algorithm `<algo>` is to be contained inside the folder named `src/<algo>`.
```

The implementation of cuML is single threaded.

## Public cuML interface
### Motivation

> The public cuML interface is stateless for two main reasons:

Reviewer 5: We should distinguish between the stateless and stateful convenience APIs here, which are also public. 

`base public API` or `core public API` seem like good choices here.

Author: Done. Also added a small terminology section to be more explicit about the terms used.

1. To ease the serialization of ML algorithm's state information (model, hyper-params, etc), enabling features such as easy pickling in the python layer.
2. To provide proper C bindings for interfacing with languages that can't consume C++ APIs  directly.

Thus, this section lays out guidelines for managing state along the API of cuML.

### General guideline

> As mentioned before, functions exposed via the base cuML C++ layer must be stateless. Things that are OK to be exposed on the interface:

Reviewer 5: I like the use of `base cuML C++ layer` here. Perhaps that wording should be used above.

Author: Yes, updated as per your previous suggestion.

1. Any [POD](https://en.wikipedia.org/wiki/Passive_data_structure) - see [std::is_pod](https://en.cppreference.com/w/cpp/types/is_pod) as a reference for C++11  POD types.
2. `cumlHandle` - since it stores GPU-related state which has nothing to do with the model/algo state. If you're working on a C-binding, use `cumlHandle_t`([reference](src/cuML_api.h)), instead.

> 3. Pointers to POD types (explicitly putting it out, even though can be considered as a POD).

Reviewer 5: Should this read `even though it can be considered as a POD`?

Author: Done.

```

An alternative correct way to expose this could be:
```cpp
// NOTE: this example assumes that TreeNode and DTParams are the model/state that need to be stored
// and passed between fit and predict methods
template <typename T> struct TreeNode { /* nested tree-like data structure, but written as a POD! */ };
struct DTParams { /* hyper-params for building DT */ };
typedef TreeNode<float> TreeNodeF;
typedef TreeNode<double> TreeNodeD;

void decisionTreeClassifierFit(const cumlHandle &handle, const float *input, int n_rows, int n_cols,
                               const int *labels, TreeNodeF *&root, DTParams params,
                               bool verbose=false);
void decisionTreeClassifierPredict(const cumlHandle &handle, const double* input, int n_rows,
                                   int n_cols, const TreeNodeD *root, int* predictions,
                                   bool verbose=false);
```
The above example understates the complexity involved with exposing a tree-like data structure across the interface! However, this example should be simple enough to drive the point across.

### Other functions on state

> This guidelines also mean that it is the responsibility of C++ API to expose methods to load and store (aka marshalling) such a data structure. Further continuing the Random Forest example,  the following methods could achieve this:

Author: Done.

> This guidelines also mean that it is the responsibility of C++ API to expose methods to load and store (aka marshalling) such a data structure. Further continuing the Random Forest example,  the following methods could achieve this:

Author: Oops.. my bad. Fixed. Thanks.

```cpp
void storeTree(const TreeNodeF *root, std::ostream &os);
void storeTree(const TreeNodeD *root, std::ostream &os);
void loadTree(TreeNodeF *&root, std::istream &is);
void loadTree(TreeNodeD *&root, std::istream &is);
```

> It is also worth noting that for algorithms such as the members of GLM, where models consist of an array of weights and therefore easy to manipulate directly by the users, such custom load/store methods might not explicitly needed.

Author: Done.

### `libcuml.so` (aka C-API wrapper over C++ API)
Following the guidelines outlined above will ease the process of "C-wrapping" the C++ API. Refer to [DBSCAN](src/dbscan/dbscan_api.h) as an example on how to properly wrap the C++ API with a C-binding. In short:
1. Use only C compatible types or objects that can be passed as opaque handles (like `cumlHandle_t`).
2. Using templates is fine if those can be instantiated from a specialized C++ function with `extern "C"` linkage.

> 3. Expose custom create/load/store/destroy methods, if the model is more complex than an array of parameters (eg: Random Forest). One possible way of working with such exposed states from the C++ layer is shown in a sample repo [here](https://github.com/teju85/managing-state-cuml).

Reviewer 2: In the linked `managing-state-cuml` repo: is `loadNodeSp` meant to load the tree node from a file? If yes, then an extra argument is needed in the `loadNodeSp`, `loadNodeDp` functions in `node.h` and `node.cpp` files.
